mayastor/src/bdev/nexus/nexus_bdev.rs

```diff
@@ -633,29 +686,140 @@ impl Nexus {
+                        subsystem.pause().await.unwrap();
+                        self.pause_state = NexusPauseState::Paused;
+                    }
+                }
             }
         }
 
+        // Keep the Nexus paused in case there are waiters.
+        if !self.pause_waiters.is_empty() {
+            let s = self.pause_waiters.pop().unwrap();
+            s.send(0).expect("Nexus pause waiter disappeared");
+        } else {
+            self.pause_state = NexusPauseState::Unpaused;
+        }
+
         Ok(())
     }
 
-    /// suspend any incoming IO to the bdev pausing the controller allows us to
+    /// Suspend any incoming IO to the bdev pausing the controller allows us to
     /// handle internal events and which is a protocol feature.
-    pub(crate) async fn pause(&self) -> Result<(), Error> {
-        if let Some(Protocol::Nvmf) = self.shared() {
-            if let Some(subsystem) = NvmfSubsystem::nqn_lookup(&self.name) {
-                subsystem.pause().await.unwrap();
+    /// In case concurrent pause requests take place, the other callers
+    /// will wait till the nexus is resumed and will continue execution
+    /// with the nexus paused once they are awakened via resume().
+    /// Note: in order to handle cuncurrent pauses properly, this function must
+    /// be called only from the master core.
+    pub(crate) async fn pause(&mut self) -> Result<(), Error> {
+        assert_eq!(Cores::current(), Cores::first());
+
+        match self.pause_state {
+            // Pause nexus if its unpaused.
+            NexusPauseState::Unpaused => {
+                self.pause_state = NexusPauseState::Pausing;
+
+                info!("{} pausing nexus", self.name);
+                if let Some(Protocol::Nvmf) = self.shared() {
+                    if let Some(subsystem) =
+                        NvmfSubsystem::nqn_lookup(&self.name)
+                    {
+                        info!(
+                            "{} pausing subsystem {}",
+                            self.name,
+                            subsystem.get_nqn()
+                        );
+                        subsystem.pause().await.unwrap();
+                        info!(
+                            "{} subsystem {} paused",
+                            self.name,
+                            subsystem.get_nqn()
+                        );
+                    }
+                }
+                self.pause_state = NexusPauseState::Paused;
+                info!("{} nexus paused", self.name);
+            }
+            // Wait till the pauser unpauses the nexus.
+            _ => {
+                info!(
+                    "{} concurrent subsystem pause detected, yielding at state: {:?}",
+                    self.name, self.pause_state,
+                );
+
+                let (s, r) = oneshot::channel::<i32>();
+                self.pause_waiters.push(s);
+
+                r.await.expect("Nexus pause sender disappeared");
```

Comment:
btw what happens on a nexus destroy, does it wait for the pause to complete? If it does not wait is it possible that while we wait here the nexus has been destroyed already which will drop all the senders?

Reply:
Removing a nexus does not impact consistency between pauser/waiters, as they all already have a valid reference to a nexus object, obtained via 'nexus_lookup()'. I presume we will have some edge cases later when implementing configuration update for a nexus that is being removed concurrently, but our pause/unpause code is unrelated to that.
So what happens is basically that all waiters that have valid references to a Nexus will get awakened, try to update the configuration (which might fail for removed nexuses) and resume the nexus, letting other waiter to repeat the same set of actions.
We definitely need to improve the way that we expose nexus instances via 'nexus_lookup()' and have some sort of reference counted Arc, but from the retire logic perspective, that's a separate thing, as we assume that as long as nexus_lookup() gives us a valid reference to a nexus, all waiters will eventually execute.

Reply:
I guess it depends on what the code that calls pause is doing right, because pause takes a self. So if you have say:
```rust
nexus.pause().await.unwrap();
nexus.do_this_that(); <- this might panic if the nexus has already been destroyed
```
Though the one I'm most wondering about is if the nexus is destroyed I think all pause_waiters will probably be woken up and might hit the expect on the r.await.
But as you said this not related to pausing, we'll need to have some sort of RFC logic to not drop the nexus if it's still being used by the pause or any other context that is still making use of the nexus.